pallets/capacity/src/types.rs

```diff
@@ -93,17 +93,32 @@ impl<T: Config> StakingAccountDetails<T> {
 	}
 
 	/// Decrease the amount of active stake by an amount and createa an UnlockChunk.
-	pub fn decrease_by(&mut self, amount: BalanceOf<T>, thaw_at: T::EpochNumber) -> DispatchResult {
-		let new_active = self.active.saturating_sub(amount);
-
-		let unlock_chunk = UnlockChunk { value: amount, thaw_at };
-
-		self.active = new_active;
+	pub fn deduct(
+		&mut self,
+		amount: BalanceOf<T>,
+		thaw_at: T::EpochNumber,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		// let's check for an early exit before doing all these calcs
+		ensure!(
+			self.unlocking.len() < T::MaxUnlockingChunks::get() as usize,
+			Error::<T>::MaxUnlockingChunksExceeded
+		);
+		let mut active = self.active.saturating_sub(amount);
+		let mut actual_unstaked: BalanceOf<T> = amount;
+
+		if active.le(&T::MinimumStakingAmount::get()) {
+			actual_unstaked = amount.saturating_add(active);
+			active = Zero::zero();
+		}
+		let unlock_chunk = UnlockChunk { value: actual_unstaked, thaw_at };
```

Comment:
This is just code golf and making sure I understand the logic. Feel free to ignore.

This has no mutation, but it is the same amount of math.

Original: sub, le, add
Below: sub, le, sub

```suggestion
		let actual_unstaked = if self.active.saturating_sub(amount).le(&T::MinimumStakingAmount::get()) {
			actual_unstaked = self.active;
		} else {
    		amount
		}
		let unlock_chunk = UnlockChunk { value: actual_unstaked, thaw_at };
		
		// self.unlocking...
		
		self.active = self.active.saturating_sub(actual_unstaked);
```

Reply:
This particular code doesn't compile, because it references itself inside the if statement. Also it does the saturating sub twice. The existing way is pretty clear and easy to read; it does the sub once, and initializes the two variables to the most likely values.